src/lib/loaders/collectorSignalsLoader.ts

```diff
@@ -0,0 +1,121 @@
+import { isFeatureFlagEnabled } from "lib/featureFlags"
+
+interface CollectorSignals {
+  bidCount?: number
+  lotWatcherCount?: number
+  partnerOffer?: { endAt: string }
+}
+
+// Compound loader that fetches signals for a given artwork using the provided context
+export const collectorSignalsLoader = async (
+  artwork,
+  ctx
+): Promise<CollectorSignals> => {
+  let bidCount, lotWatcherCount, partnerOffer
+
+  const artworkId = artwork._id
+
+  const isInSale = artwork.sale_ids?.length > 0
+
+  const unleashContext = {
+    userId: ctx.userID,
+  }
+  const partnerOfferCollectorSignalsEnabled = isFeatureFlagEnabled(
+    "emerald_signals-partner-offers",
+    unleashContext
+  )
+  const auctionsCollectorSignalsEnabled = isFeatureFlagEnabled(
+    "emerald_signals-auction-improvements",
+    unleashContext
+  )
+
+  // Handle signals for auction artworks
+  if (isInSale && auctionsCollectorSignalsEnabled) {
+    const activeSaleArtwork = await getActiveSaleArtwork(
+      {
+        artworkId,
+        saleIds: artwork.sale_ids,
+      },
+      ctx
+    )
+
+    if (activeSaleArtwork) {
+      if (artwork.recent_saves_count) {
+        lotWatcherCount = artwork.recent_saves_count
+      }
+      if (activeSaleArtwork.bidder_positions_count) {
+        bidCount = activeSaleArtwork.bidder_positions_count
+      }
+    }
+  }
+
+  // Handle signals for non-auction artworks
+  if (artwork.purchasable && partnerOfferCollectorSignalsEnabled) {
+    if (ctx.mePartnerOffersLoader) {
+      try {
+        const partnerOffers = await ctx.mePartnerOffersLoader({
+          artwork_id: artworkId,
+          sort: "-created_at",
+          size: 1,
+        })
+
+        partnerOffer = partnerOffers.body[0]
+      } catch (error) {
+        console.error(
+          "collectorSignalsLoader: Error fetching partner offers",
+          error
+        )
+      }
+    }
+  }
```

Comment:
@mzikherman - Regarding your comment https://github.com/artsy/metaphysics/pull/5843#discussion_r1687206253:

> I actually want to caution against zealous catch. In MP, without any catch - you get pretty desirable behavior. There is Sentry reporting, the error shows up in errors in the response and the corresponding field is null, and the error in the response includes relevant information to help debug.
> 
> There are a lot of examples of catch that simply log 'Error fetching .... ' -> not actually providing any information, and the resulting field is usually still null, thus catch does nothing there. Or catch and throwing your own error there - more or less the same thing.
> 
> Basically you shouldn't need a catch in most cases.

I took this comment to be in reference to this field & so removed most catches in this file in 78b2a6c7d43f60d1d3c4679a31e0a6cf5f0d660e, but I wonder if this could cause entire artwork connections to be null out as it is part of the initial resolver. If I misinterpreted let me know and I'll put them back.

Reply:
Try it and see! Put a `throw` in there and see what happens to the field/resolutions. This kind-of a thing, where you don't want the whole connection to `null` out (just the `collectorSignals` part of the `node`) _is_ totally something where you may want to `catch`.

Reply:
Confirmed this breaks the entire parent resolution, so put it back. I think this is good to merge 👊